dask/array/linalg.py

```diff
@@ -0,0 +1,125 @@
+"""
+   Copyright (c) 2015, Mariano Tepper, Duke University.
+   All rights reserved.
+
+   This file is part of dask and is under the BSD 3-Clause License,
+   which can be found at http://opensource.org/licenses/BSD-3-Clause
+"""
+
+import numpy as np
+from itertools import count, product
+import dask.array as da
+from math import ceil
+import operator
+
+names = ('tsqr_%d' % i for i in count(1))
+
+
+def _findnumblocks(shape, blockshape):
+    def div_ceil(t):
+        return int(ceil(float(t[0]) / t[1]))
+
+    nb = [div_ceil(elem) for elem in zip(*[shape, blockshape])]
+    return tuple(nb)
+
+
+def qr(data, name=None):
+    """
+    Implementation of the direct TSQR, as presented in:
+
+    A. Benson, D. Gleich, and J. Demmel.
+    Direct QR factorizations for tall-and-skinny matrices in
+    MapReduce architectures.
+    IEEE International Conference on Big Data, 2013.
+    http://arxiv.org/abs/1301.1071
+
+    :param data: dask array object
+    Shape of the blocks that will be used to compute
+    the blocked QR decomposition. We have the restrictions:
+    - blockshape[1] == data.shape[1]
+    - blockshape[0]*data.shape[1] must fit in main memory
+    :return: tuple of dask.array.Array
+    First and second tuple elements correspond to Q and R, of the
+    QR decomposition.
+    """
+    if not (data.ndim == 2 and                    # Is a matrix
+            len(data.blockdims[1]) == 1):         # Only one column block
+        raise ValueError(
+            "Input must have the following properites:\n"
+            "  1. Have two dimensions\n"
+            "  2. Have only one column of blocks")
+    blockshape = (data.blockdims[0][0], data.blockdims[1][0])
+    m, n = data.shape
+
+    prefix = name or next(names)
+    prefix += '_'
+
+    numblocks = _findnumblocks(data.shape, blockshape)
+
+    name_qr_st1 = prefix + 'QR_st1'
+    dsk_qr_st1 = da.core.top(np.linalg.qr, name_qr_st1, 'ij', data.name, 'ij',
+                             numblocks={data.name: numblocks})
+    # qr[0]
+    name_q_st1 = prefix + 'Q_st1'
+    dsk_q_st1 = {(name_q_st1, i, 0): (operator.getitem, (name_qr_st1, i, 0), 0)
+                 for i in xrange(numblocks[0])}
+    # qr[1]
+    name_r_st1 = prefix + 'R_st1'
+    dsk_r_st1 = {(name_r_st1, i, 0): (operator.getitem, (name_qr_st1, i, 0), 1)
+                 for i in xrange(numblocks[0])}
+
+    # Stacking for in-core QR computation
+    def _vstack(*args):
+        tup = tuple(args)
+        return np.vstack(tup)
```

Comment:
Ah, I see.  `np.vstack` didn't like the iterator that dask gave to it.  If we're comfortable using nested tasks then we could resolve this by creating tasks like the following:

* Before:  `(_vstack, [a, b, c])`
* After: `(np.vstack, (tuple, [a, b, c]))`